src/main/java/com/hierynomus/sshj/transport/cipher/ChachaPolyCipher.java

```diff
@@ -0,0 +1,164 @@
+/*
+ * Copyright (C)2009 - SSHJ Contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.hierynomus.sshj.transport.cipher;
+
+import java.nio.ByteBuffer;
+import java.security.GeneralSecurityException;
+import java.security.spec.AlgorithmParameterSpec;
+import java.util.Arrays;
+import javax.crypto.spec.IvParameterSpec;
+
+import net.schmizz.sshj.common.SSHRuntimeException;
+import net.schmizz.sshj.common.SecurityUtils;
+import net.schmizz.sshj.transport.cipher.BaseCipher;
+
+public class ChachaPolyCipher extends BaseCipher {
+
+    private static final int CHACHA_KEY_SIZE = 32;
+    private static final int AAD_LENGTH = 4;
+    private static final int POLY_TAG_LENGTH = 16;
+
+    private static final String CIPHER_CHACHA = "CHACHA";
+    private static final String MAC_POLY1305 = "POLY1305";
+
+    private static final ByteBuffer LONG_TO_BYTES_BUFFER = ByteBuffer.allocate(8);
+    private static final byte[] POLY_KEY_INPUT = new byte[32];
+
+    private final int authSize;
+
+    private byte[] encryptedAad;
+
+    protected Mode mode;
+    protected javax.crypto.Cipher aadCipher;
+    protected javax.crypto.Mac mac;
+    protected java.security.Key cipherKey;
+    protected java.security.Key aadCipherKey;
+
+    public ChachaPolyCipher(int authSize, int bsize, String algorithm) {
+        super(0, bsize, algorithm, CIPHER_CHACHA);
+        this.authSize = authSize;
+    }
+
+    @Override
+    public int getAuthenticationTagSize() {
+        return authSize;
+    }
+
+    @Override
+    public void init(Mode mode, byte[] key, byte[] iv) {
+        this.mode = mode;
+
+        cipherKey = getKeySpec(Arrays.copyOfRange(key, 0, CHACHA_KEY_SIZE));
+        aadCipherKey = getKeySpec(Arrays.copyOfRange(key, CHACHA_KEY_SIZE, 2 * CHACHA_KEY_SIZE));
+        try {
+            cipher = SecurityUtils.getCipher(CIPHER_CHACHA);
+            aadCipher = SecurityUtils.getCipher(CIPHER_CHACHA);
+            mac = SecurityUtils.getMAC(MAC_POLY1305);
+        } catch (GeneralSecurityException e) {
+            cipher = null;
+            aadCipher = null;
+            mac = null;
+            throw new SSHRuntimeException(e);
+        }
+
+        setSequenceNumber(0);
+    }
+
+    @Override
+    public void setSequenceNumber(long seq) {
+        byte[] seqAsBytes = LONG_TO_BYTES_BUFFER.putLong(0, seq).array();
```

Comment:
`LONG_TO_BYTES_BUFFER` is static. If there are two concurrent `setSequenceNumber` from different instances are called simultaneously, one of them can get wrong results.

Reply:
Haha, you beat me to it ;) Indeed, there is a big concurrency issue here. Prevent static mutable fields such as buffers etc.